File: 0300-0399/0384. Shuffle an Array/0384. Shuffle an Array.js
/**
 * @param {number[]} nums
 */
var Solution = function (nums) {
  this.nums = nums;
};

/**
 * Resets the array to its original configuration and return it.
 * @return {number[]}
 */
Solution.prototype.reset = function () {
  return this.nums;
};

/**
 * Returns a random shuffling of the array.
 * @return {number[]}
 */
Solution.prototype.shuffle = function () {
  const result = this.nums.slice();
  for (let i = 1; i < result.length; i++) { //必须要从1开始  因为从0开始 第一个随机数无论怎么生成都是0
    const pos = Math.trunc(Math.random() * (i + 1));  // 取[0,i]的随机数
    [result[i], result[pos]] = [result[pos], result[i]];
  }
  return result;
};
  /**
* Your Solution object will be instantiated and called as such:
* var obj = new Solution(nums)
* var param_1 = obj.reset()
* var param_2 = obj.shuffle()
*/

File: 0300-0399/0384. Shuffle an Array/0384. Shuffle an Array.py
import random
from typing import List


class Solution:
    def __init__(self, nums: List[int]):
        self.nums = nums

    def reset(self) -> List[int]:
        """
        Resets the array to its original configuration and return it.
        """
        return self.nums

    def shuffle(self) -> List[int]:
        """
        Returns a random shuffling of the array.
        """
        copy = self.nums[:]
        for i in range(1, len(copy)):
            # randrange(0,n)   a<n a>=0
            # randint(0,n)   a<=n  a>=0
            pos = random.randrange(0, i + 1)
            copy[i], copy[pos] = copy[pos], copy[i]
        return copy


# Your Solution object will be instantiated and called as such:
# obj = Solution(nums)
# param_1 = obj.reset()
# param_2 = obj.shuffle()
